elements/ip/ipfilter.cc

```diff
@@ -1266,19 +1296,41 @@ IPFilter::parse_program(Classification::Wordwise::CompressedProgram &zprog,
 	}
 
 	prog.finish_subtree(tree, Classification::c_and, -slot);
+	progs.push_back(prog);
     }
 
-    if (tree.size())
-	prog.finish_subtree(tree, Classification::c_or, Classification::j_never, Classification::j_never);
-
-    // click_chatter("%s", prog.unparse().c_str());
-    static const int offset_map[] = { offset_net + 8, offset_net + 3 };
-    prog.optimize(offset_map, offset_map + 2, Classification::offset_max);
+    int num_progs = progs.size();
+    if (num_progs == 1) {
+        progs[0].optimize(offset_map, offset_map + 2, Classification::offset_max);
+    } else {
+        //recursivly merge programs in pairs of two
+        int n = 1;
+        while (n * 2 <= num_progs) {
+            for (int i = 0; i < num_progs; i += n * 2) {
+                int a = i;
+                int b = i + n;
+                if (b < num_progs) {
+                    merge_programs(progs[a], progs[b]);
+                    progs[a].optimize(offset_map, offset_map + 2, Classification::offset_max);
+                } else {
+                    //odd prog at end, not merging yet
+                    continue;
+                }
+            }
+            n = n * 2;
+        }
+
+        //check if there is an odd program at the end not yet merged
+        if (n != num_progs) {
+            merge_programs(progs[0], progs[n]);
+            progs[0].optimize(offset_map, offset_map + 2, Classification::offset_max);
+        }
+    }
```

Comment:
fails to route final `j_failure` branches to `j_never`, which is required for configs like `IPFilter(0 ip src 10.0.0.1)`